### ISSUE TITLE
Block structure for extensibility

### ISSUE THREAD

**Issue by dinhtungdu** [woothemes/woocommerce#42178]
We want the new filter block be easy to use for merchants/shoppers and be easy to extend for developer at the same time. We're already experimenting some [options](https://github.com/woocommerce/woocommerce-blocks/discussions/11141#discussioncomment-7208775) and we expect to decide to use one within this cycle.

Edit (Mar 1st, 2024): the new proposal is up on https://github.com/woocommerce/woocommerce/discussions/45154

**Comment by dinhtungdu** [comment]
After a long discussion and some experimentation with @luisherranz suggestions, I ended up with, IMO, the best solution for structuring filter blocks for both extensibility and usability. The structure is being implemented in woocommerce/woocommerce-blocks#11191: 

- `Collection Filters` block acts as the wrapper block providing context:
  - It detects the inner filter blocks to build queries to fetch Store API/ Product Collection Data. We do this to fetch only necessary collection data.
  - It fetches the Collection Data and prepares the Interactivity API state then passes it down to filter blocks through a context named `filterData`. By doing so, we only fetch and process the collection data once, thus we're improving the performance and removing the duplication.
  - It enqueues the Interactivity API Store for all filter blocks.
- Filter blocks as the view blocks which consume the `filterData` context:
  - View block is the new term to describe the new filter blocks that don't need to do any data fetching. Instead, they just consume the filter data from the block context.
  - Filter blocks have two requirements:
    - Have `woocommerce/collection-filters` as the `ancestor`.
    - Have an attribute named `filterType`, with the type set to `string` and default value to one of the four values: `price`, `rating`, `attribute`, `rating`.
  - Filter blocks can define and enqueue their own Interactivity API state/action if needed.
  - Filter blocks can have their own attributes if needed.

By following this structure, we're delivering a good experience for both merchants and developers:
- Usability: Unlike the option introduced in [the proposal](https://github.com/woocommerce/woocommerce-blocks/discussions/11141), which turns filter blocks into parent blocks, the block complexity is unchanged.
- Extensibility: Developers who want to change how filter blocks look can create new blocks to implement their representation. The learning curve is minimal as we aren't introducing any new 'things'. Block developers are familiar with block, ancestor, and context.

**Comment by dinhtungdu** [comment]
@samueljseay raised a good point on the [current suggested structure](https://github.com/woocommerce/woocommerce-blocks/pull/11191) that the Collection Filters block is bloated as it's responsible for too many things:
1. Detect filter types from inner blocks.
2. Fetch collection data.
3. Process the Interactivity API store data.
4. Enqueue the Interactivity API store.
5. Pass the store data down to inner blocks through context.

Among these 'things', the 1, 2, and 5 are must-haves for the Collection Filters block. The 3 and 4 can be delegated to filter blocks, but I did it for a reason. Let's examine if the reason is a good enough trade-off:

> We process, enqueue, and pass the Interactivity API in the Collection Filters block so inner filter blocks don't have to care about processing and enqueue the store data. Given there are two price filter blocks: price range and price slider, by doing like we're doing, the price filter blocks just need to implement the view/representation without care about the **same** data fetching/enqueueing logic. They're free to do so if they have additional data but we're providing the foundation so developers don't have to repeat.

At this point, I can already see that we're saving developers' effort but, at the same time, we're limiting what they can do.

**Comment by dinhtungdu** [comment]
Passing `collectionData` down through block context might be a quick way to solve the limitation.

**Comment by dinhtungdu** [comment]
Currently, in https://github.com/woocommerce/woocommerce-blocks/pull/11191, Price Filter block defines a block attribute named `filterType`, Collection Filters block uses that attribute to build the final query for fetching product collection data (attribute count, price range...) from the Store API. To make it easier to write the filter blocks, the Collection Filters block also transforms the collection data into an Interactivity API store, so filter blocks don't have to fetch any data (that's why I called them view block). This works well for the Price Filter but creates an issue for the Attribute Filter block.

Attribute Filter block doesn't only need the store data, but also the term data. This leads to two options:
- Collection Filters block fetches the term data and passes it down: We will need the Attribute Filter block to define another block attribute, Collection Filters block needs another logic to fetch term data.
- We fetch the terms inside the Attribute Filter block, violating the view block definition, and potentially making third-party developers confused.

With these options in mind, we can either:
- Abandon the current direction, and only pass the product collection data down without processing (like we're doing on the `trunk`) => this is more straightforward and clearer, hence easier for third-party developers to grasp, but the filter blocks are more complex.
- Continue transforming the product collection data in the Collection Filters block, fetching additional data in filter blocks. This makes filter block implementation simpler and easier to maintain but can be confusing.

**Comment by samueljseay** [comment]
@dinhtungdu 

One thing I was thinking about is whether separating out data processing work into functions (or hooks) and code outside the collections block reduces complexity inside the block enough. The end result is sort of the same, the collections block is ultimately responsible for passing down already parsed data, but with everything being separated out into functions with their own supporting unit tests it takes the complexity out of the collection filter block itself. A little JSDoc and some separation goes a long way.


Then your view block pattern can still be done.

It also means if later we want to take the processing out of collection block and put it elsewhere its trivial to do that.
Does that reduce complexity enough you think?  Or are there other complexity issues here?

I think we don't want collection block to be tightly coupled to the logic anyway, because it makes it harder to refactor later.

**Comment by dinhtungdu** [comment]
As discussed on Slack, I'll update woocommerce/woocommerce-blocks#11191 to move the data processing logic out of Collection Filters block to relevant filter block, and make sure third party dev can reuse those logic to implement their own blocks.

**Comment by dinhtungdu** [comment]
After examining the work in woocommerce/woocommerce-blocks#11191, we (@dinhtungdu and @samueljseay) concludes that's not the good direction to move forward:

- woocommerce/woocommerce-blocks#11191 makes the `Collection Filters` block really bloated.
- We wanted to make filter blocks only consume the data passed from `Collection Filters`, not fetching data. But this is impossible, filter blocks like Attribute Filter needs to fetch its own data => this make developers confused by having data fetching logic in two places.

Next steps are:
- Make the utilities/functions available for third party developer.
- Revisit woocommerce/woocommerce-blocks#11139.

**Comment by luisherranz** [comment]
> Make the utilities/functions available for third party developer.

I agree this would be a good solution in terms of simplicity and flexibility. Please, keep reporting your findings 🙂

**Comment by dinhtungdu** [comment]
FYI, I created a new proposal in https://github.com/woocommerce/woocommerce/discussions/45154

**Comment by kkoteen** [comment]
Hi @dinhtungdu - can you update next steps or close this issue if it's no longer valid? Thanks!

**Comment by dinhtungdu** [comment]
Closing this issue as we implemented the proposal.
